### ISSUE TITLE
proposal: default to literal searches with keyword, regex, etc. fallbacks on no results

### ISSUE THREAD

**Issue by ijt** [sourcegraph/sourcegraph#4447]
We want to support simple, literal queries but also make it easy for regex queries to continue working. Here is a proposal to accomplish that.

Goals
-----
1. Literal searches like `NewRouter(` just work.
2. Searches can be shared by URL and preferably also by query.
3. Old search URLs continue to produce the same results as before.
4. Old regex-based search queries continue to produce the same results as before.
5. Saved searches from before continue to work unchanged.
6. New saved searches work smoothly.
7. Scripts calling the `src` CLI continue to work unchanged.
8. Latency is not noticeably increased.

Proposal
--------
We start by assuming nothing since there is plenty of uncertainty about what sort of query will be made when the user lands on the search page. Even if the previous query was meant as a regular expression, we don't know if this one will be. The previous search may have yielded results with case sensitive matching, but this search may need to be insensitive.

The user can tell Sourcegraph what kind of search to run by adding a new flag to the query called `exp`, short for "expression" as in "regular expression".  This flag can have the following values:
* `literal` or `lit`: the query is interpreted literally
* `keyword` or `key`: like literal but spaces are converted to `.*` regexes
* `regexp` or `regex` or `rx`: the query is interpreted as a regular expression
* `keyregexp` or `keyregex` or `krx`: the query is interpreted as a regular expression, except that spaces are converted to `.*`. This is how Sourcegraph search currently works, and would be a sensible default for the GQL API `Search` query for backwards compatibility.
* `structural` or `sx` or `comby`: the query is interpreted as one of Rijnard's patterns with holes like `:[_]` or `:[foo]` (future work)
* `unknown`: the query is tried as each of `literal`, `keyword`, `regexp`, `keyregexp`, `structural` until some results are found, and an additional field is set in the GQL result saying which type. The match type is displayed in the UI. This would make sense as the default for the web app, probably implemented by setting a new `ExpDefault` field in the GQL `Search` query to `unknown` instead of its default value of `keyregexp`.

Saved queries will need updating to work with this. The `saved_queries` table will have a new column called `exp_default` whose default value would be `keyregexp` to ensure backwards compatibility for existing saved queries. All new saved queries will set `exp_default` to `unknown`.

Let's have a look at how well this proposal accomplishes the goals:

1. Literal searches like `NewRouter(` just work.
	* They do because the default `unknown` search type first tries literal search.
2. Searches can be shared by URL and preferably also by query.
	* The searches are just as shareable as before, by URL and by query.
3. Old search URLs continue to produce the same results as before.
	* Usually they will. Ones that get unintended literal search results will need to have `exp:keyregexp` to continue working.
4. Old regex-based search queries continue to produce the same results as before.
	* Same as 3.
5. Saved searches from before continue to work unchanged.
	* The new `saved_queries` column of `exp_default` defaulting to `keyregexp` handles this.
6. New saved searches work smoothly.
	* Saving new queries with `exp_default` set to `unknown` handles this.
7. Scripts calling the `src` CLI continue to work unchanged.
	* This is done by having the default for `exp` be `keyregexp` outside of the web app.
8. Latency is not noticeably increased.
	* On my local dev instance with 65 repos, empty results are taking 20-30ms, so it's at least plausible that `exp:unknown` won't make regex searches noticeably worse. We may find that it works best to run the different possible interpretations in parallel since most of the time most of them will be ruled out or return nothing.

Optimizations
--------------
If `unknown` is selected, `regexp` and `keyregexp` can be skipped if the query contains no special regex characters or is not a valid regex. Also `structural` can be skipped if the query contains no holes like `:[foo]`.

~One way to get everything but `structural` at once would be to make a big regex with a disjunction (or) of all the possible interpretations. For example, given a query like `func ma[ik][ne]` the big regex would be `(("func ma[ik][ne]")|("func".*"ma[ik][ne]")|(func ma[ik][ne])|(func.*ma[ik][ne]))`. It might be faster. We'd have to run some benchmarks and find out. The `rx` and `krx` groups would have to be excluded if they don't compile. It would also be less clear how to communicate to the user what type of match has been found.~ Making one big regexp won't work because it would mix search results from different interpretations of the query.

Alternatives
------------

### [Always try everything](https://github.com/sourcegraph/sourcegraph/issues/4178)
The proposal here can be considered a refinement of this earlier proposal, with the addition of the `exp` flag and the change that the `unknown` value returns results for the first search type that has them instead of results for all possible interpretations. It was too confusing to consider how to mix together search results from different interpretations of the query.

### [Make regexp opt-in via UI toggle button](https://github.com/sourcegraph/sourcegraph/issues/4438)
This is a practical near-term proposal, however it has some drawbacks compared to the fallback proposal here:

* Query sharing is complicated by the need to click UI widgets when Alice and Bob have different settings.
* As noted earlier, we often want to do a literal search right after having done a regex search and it's nice not to have to fiddle with the UI if we don't have to.
* If the `.*` box is checked and Alice inputs `foo+bar` while forgetting that `+` is a regex special character, she'll be surprised at the results. With the `exp:unknown` default of this proposal, `foo+bar` is more likely to work as intended, probably getting literal results and stopping there.

**Comment by keegancsmith** [comment]
I like this idea a lot :+1:

How do we communicate to users who are trying to write a regex, but the regex is invalid? Right now it seems would just show no results (excluding the case where the invalid regex matches a literal).

How does this translate to other parts where we use regex. EG repo and file matching?

**Comment by ijt** [comment]
> How do we communicate to users who are trying to write a regex, but the regex is invalid? Right now it seems would just show no results (excluding the case where the invalid regex matches a literal).
The `SearchResult` GQL type could contain a `regexpErr` field containing any error in parsing the query as a regex.

The UI could say something like "Showing literal results. The query has an error as a regexp: $regexpErr". 

> How does this translate to other parts where we use regex. EG repo and file matching?

I'd like to leave that beyond the scope of this proposal, since it would mean at least doubling the number of things we have to try. Most likely we could get more mileage there with a "did you mean?" feature since it would be a lot easier to detect common mistakes in repo and file patterns than in patterns for general code.

**Comment by felixfbecker** [comment]
I'd like to say that having **five** different search modes exposed through an operator, and in the case it is not specified, falling back to one of them, sounds overly complex and confusing to me. It may be deterministic, but it may not feel like that for the user who doesn't understand the fallback logic. The results then appear random, reducing trust. I would classify this proposal as "magic behaviour", which can be awesome sometimes, but imo usually is more confusing than convenient.

To the toggle feedback:

> - Query sharing is complicated by the need to click UI widgets when Alice and Bob have different settings.

Not if we intercept copy events or Alice shares the URL.

> - As noted earlier, we often want to do a literal search right after having done a regex search and it's nice not to have to fiddle with the UI if we don't have to.

You don't necessarily have to fiddle with the UI. The operator can be added to the 	query, but there would of course also be keyboard shortcuts for these, just like in the editor. These are actually faster and more discoverable than an operator.

**Comment by ijt** [comment]
> I'd like to say that having five different search modes exposed through an operator, and in the case it is not specified, falling back to one of them, sounds overly complex and confusing to me. It may be deterministic, but it may not feel like that for the user who doesn't understand the fallback logic. The results then appear random, reducing trust. I would classify this proposal as "magic behaviour", which can be awesome sometimes, but imo usually is more confusing than convenient.

Confusion could happen in a situation where the user wants to use a regex as the query, but that regex is present literally somewhere in the code being searched. I think that (probably rare) case can be handled well by detecting that it's a valid regex with special characters and putting a "did you mean" link with an `exp:rx` field above the search results.

Another possible source of confusion might be if there are matches with query spaces treated literally, whereas the user really wanted all the matches that could be found by converting the query spaces to `.*`. I doubt this would be a common problem, but even if it is then we could put a notice on the search results saying spaces have been treated literally and adding a link to expand them to `.*`. Probably we'd want to have two search fields to support this: `exp:[?|lit|rx|sx]` and `space:[?|lit|.*]`. 

> Not if we intercept copy events or Alice shares the URL.

I guess the idea is to somehow inject a field into the query as it is being copied, based on the `.*` toggle state. But then it's unclear what to do when someone tries to copy a substring of the query.

> You don't necessarily have to fiddle with the UI. The operator can be added to the query, but there would of course also be keyboard shortcuts for these, just like in the editor. These are actually faster and more discoverable than an operator.

What I'm proposing doesn't rule out having UI widgets. For example, there could be a radio button set like `[?|lit|.*]` where `?` for `unknown` is set by default. They could two-way sync with the `exp:` field in the search box.

**Comment by felixfbecker** [comment]
That's not the confusion I meant - what I meant is that the search strategy depends on the _query_ (not the syntax) and even the _result_ (i.e. state of repos). That means a user may do a regex query, and notice it interpreted the regex characters. But then on the next query the user tries, Sourcegraph interprets the regex characters literally, because there were literal results. That is confusing behavior. Same for the other strategies

**Comment by ijt** [comment]
> But then on the next query the user tries, Sourcegraph interprets the regex characters literally, because there were literal results. That is confusing behavior.

In this situation, we can display text above the search results saying "Results for query as literal: `thing[12]` chosen by [fallback rules](fake). [Try query again as regex.](cantclickthis)"

It may be a little surprising but I don't see how this would be confusing. We offer a clear explanation and a simple action to take to get the other interpretation.

There can be site-wide configuration like `queryExpDefault` to set the default for `exp` to `lit` or `rx` in case there is too much confusion.

**Comment by keegancsmith** [comment]
@attfarhan A version of this is done now right? We can close this?

**Comment by slimsag** [comment]
Fixed by https://github.com/sourcegraph/sourcegraph/pull/5899
